services/audit-service/src/main/java/datawave/microservice/audit/auditors/file/FileAuditor.java

```diff
@@ -0,0 +1,220 @@
+        config.set("fs.hdfs.impl", org.apache.hadoop.hdfs.DistributedFileSystem.class.getName());
+        config.set("fs.file.impl", org.apache.hadoop.fs.LocalFileSystem.class.getName());
+        
+        if (builder.configResources != null) {
+            for (String resource : builder.configResources)
+                config.addResource(new Path(resource));
+        }
+        
+        if (builder.fileUri != null)
+            fileSystem = FileSystem.get(new URI(builder.fileUri), config);
+        else
+            fileSystem = FileSystem.get(config);
+        
+        basePath = new Path(builder.path);
+        if (!fileSystem.exists(basePath))
+            fileSystem.mkdirs(basePath);
+        
+        this.sdf = new SimpleDateFormat("'" + builder.prefix + "'-yyyyMMdd_HHmmss.SSS'.json'");
+    }
+    
+    @Override
+    public void audit(AuditParameters auditParameters) throws Exception {
+        
+        // convert the messages to URL-encoded JSON
+        Map<String,String> auditParamsMap = auditParameters.toMap();
+        auditParamsMap.forEach((key, value) -> auditParamsMap.put(key, urlEncodeString(value)));
+        String jsonAuditParams = mapper.writeValueAsString(auditParamsMap) + "\n";
+        
+        writeLock.lock();
+        try {
+            // if the file/stream is null, doesn't exist, or the file is too old/big, create a new file & output stream
+            if (currentFile == null || !fileSystem.exists(currentFile) || isFileTooOld() || isFileTooBig())
+                createNewFile();
+            
+            writeAudit(jsonAuditParams);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+    
+    protected String urlEncodeString(String value) {
+        try {
+            return URLEncoder.encode(value, "UTF8");
+        } catch (UnsupportedEncodingException e) {
+            log.error("Unable to URL encode value: " + value);
+        }
+        return value;
+    }
+    
+    protected void writeAudit(String jsonAuditParams) throws Exception {
+        OutputStream appendStream = (fileSystem instanceof LocalFileSystem) ? new FileOutputStream(currentFile.toString(), true)
+                        : fileSystem.append(currentFile);
+        appendStream.write(jsonAuditParams.getBytes("UTF8"));
+        appendStream.close();
+    }
+    
+    protected void createNewFile() throws IOException, ParseException {
+        // create a new file and output stream
+        Date currentDate = null;
+        do {
+            // if this isn't our first attempt, sleep for a short time
+            if (currentDate != null) {
+                try {
+                    Thread.sleep((long) (Math.random() * 10));
+                } catch (InterruptedException e) {
+                    // getting interrupted is ok
+                }
+            }
+            
+            currentDate = new Date();
+            currentFile = new Path(basePath, sdf.format(currentDate));
```

Comment:
Can we use podname or hostname instead to avoid the possibility of conflicts?  The only reason I say this is hdfs can be troublesome as a locking mechanism.

Reply:
fixed in next commit